File: Python/Data Structures/Stack.py
"""
! Stack 
stack is a data structure that follows the LIFO (Last In First Out) and (first in last out) principle. 
in this data stucture you can add data to a stack but whatever you add next will go above what you added before it 
this is why the first thing in will be the last thing out as everything after it will have to be removed before the first element can be removed
adding to a stack is called push and removing from a stack is called pop
NOTE you can only add and remove from one place in a stack which is the top of the stack
meaning the frist data you entered will be at the bottom and the last data you entered will be at the top
stacks are useful for undo and redo functions in applications, its NOT built in 

Operations with stack
1. push() - adds an element to the stack
2. pop() - removes an element from the stack
3. peek() - returns the top element from the stack
4. isEmpty() - returns true if the stack is empty

Methods of implementation
1. array/list implementation
2. modules
"""

# # method 1 using list here push will append and pop will pop from list both operation done at the end of the list
# stack = [] # making the list for stack
# # addint to stack
# stack.append(10)
# stack.append(20)
# stack.append(30)
# #removing from stack NOTE: pop a python built in will remove the last element added from a list
# stack.pop() 
# stack.pop()
# stack.pop() # last pop 

# # to check len or if empty
# print(len(stack)) # return  s 0, len is a built in and gives the number of elements in a list

# # to get elements
# stack.append(10)
# stack.append(20)
# # to accsess the last element added use index -1 or len (stack)-1
# print(stack[-1]) # returns 20

# ! Now lets combine this into a object oriented program with methods
# NOTE we can introduce a limit to the stack array using user input then cheking in the push operation the len of stack
stack = []
def push():
    item = input("Emter Element to push: ")
    stack.append(item)
    print(f"Item {item} pushed to stack")
def pop():
    if not stack: # checks if stack is empty
        print("Stack is empty")
    else:
        e = stack.pop()
        print("removed element: ",e)
def peak():
    if not stack:
        print("Stack is empty")
    else:
        print(stack[-1])
def printStack():
    print(stack)

# can add other methods

# to allow user input
while True:
    print("1. Push")
    print("2. Pop")
    print("3. Peek")
    print("4. Show Stack")
    print("5. quit")
    choice = int(input())
    if choice == 1:
        push()
    elif choice == 2:
        pop()
    elif choice == 3:
        peak()
    elif choice == 4:
        printStack()
    elif choice == 5:
        break
    else:
        print("Invalid choice")

File: Java/Data_Structures/Stack.java
package Java.Data_Structures;

import java.util.ArrayList;
import java.util.Scanner;

/**
 * Stack
 * A stack is a data structure that follows the LIFO (Last In First Out) principle.
 * In this data structure, you can add data to a stack but whatever you add next will go above what you added before it.
 * This is why the first thing in will be the last thing out, as everything after it will have to be removed before the first element can be removed.
 * Adding to a stack is called push and removing from a stack is called pop.
 * 
 * NOTE: You can only add and remove from one place in a stack which is the top of the stack.
 * This means the first data you entered will be at the bottom and the last data you entered will be at the top.
 * Stacks are useful for undo and redo functions in applications.
 * 
 * Operations with stack:
 * 1. push() - adds an element to the stack
 * 2. pop() - removes an element from the stack
 * 3. peek() - returns the top element from the stack
 * 4. isEmpty() - returns true if the stack is empty
 * 
 * Methods of implementation:
 * 1. ArrayList implementation (shown here)
 * 2. LinkedList implementation (discussed at the bottom)
 * 3. Using Java's built-in Stack class
 */
public class Stack<T> {
    private ArrayList<T> stack; // Using generic type T instead of Object for better type safety

    // Constructor
    public Stack() {
        stack = new ArrayList<>();
    }

    // Push operation - adds element to the top of the stack
    public void push(T item) {
        stack.add(item);
        System.out.println("Item " + item + " pushed to stack");
    }

    // Pop operation - removes element from the top of the stack
    public T pop() {
        if (isEmpty()) {
            System.out.println("Stack is empty");
            return null;
        } else {
            T item = stack.remove(stack.size() - 1);
            System.out.println("Removed element: " + item);
            return item;
        }
    }

    // Peek operation - returns the top element without removing it
    public T peek() {
        if (isEmpty()) {
            System.out.println("Stack is empty");
            return null;
        } else {
            return stack.get(stack.size() - 1);
        }
    }

    // Check if stack is empty
    public boolean isEmpty() {
        return stack.size() == 0;
    }

    // Get size of the stack
    public int size() {
        return stack.size();
    }

    // Display all elements in the stack
    public void printStack() {
        System.out.println(stack);
    }

    // Main method for testing the stack implementation
    public static void main(String[] args) {
        Stack<String> stack = new Stack<>(); // Creating a Stack of Strings for demonstration
        Scanner scanner = new Scanner(System.in);
        boolean running = true;

        while (running) {
            System.out.println("1. Push");
            System.out.println("2. Pop");
            System.out.println("3. Peek");
            System.out.println("4. Show Stack");
            System.out.println("5. Quit");
            System.out.print("Enter your choice: ");
            
            int choice = scanner.nextInt();
            scanner.nextLine(); // Clear the input buffer
            
            switch (choice) {
                case 1:
                    System.out.print("Enter Element to push: ");
                    String item = scanner.nextLine();
                    stack.push(item);
                    break;
                case 2:
                    stack.pop();
                    break;
                case 3:
                    System.out.println("Top element: " + stack.peek());
                    break;
                case 4:
                    stack.printStack();
                    break;
                case 5:
                    running = false;
                    break;
                default:
                    System.out.println("Invalid choice");
            }
        }
        scanner.close();
    }
}

/**
 * LinkedList Implementation of Stack
 * 
 * Another way to implement a Stack is using a LinkedList:
 * - In a LinkedList implementation, each element points to the next element
 * - The head of the LinkedList would represent the top of the stack
 * - Push operation: Create a new node and make it the new head
 * - Pop operation: Remove the head node and update the head to the next node
 * - Peek operation: Return the value of the head node without removing it
 * 
 * Advantages of LinkedList implementation:
 * - Dynamic size (no need to resize)
 * - Efficient memory utilization (only allocates memory when needed)
 * - Constant time O(1) for push and pop operations
 * 
 * Disadvantages:
 * - Extra memory needed for storing references/links
 * - No random access like ArrayList
 * 
 * Implementation would involve:
 * 1. Creating a Node class with data and next pointer
 * 2. Maintaining a top/head reference
 * 3. For push: Create new node, link to current top, update top
 * 4. For pop: Save data from top, move top to next node, return saved data
 */
